### PR TITLE
improve noop elimination for view

### PR DESCRIPTION
**Pull request by BoyuanFeng**
This PR improves noop elimination.

### View Noop

```python
>>> torch.Size([1,2,3]) == [1,2,3]
False
>>> torch.Size([1,2,3]) == (1,2,3)
True
```
So we add `tuple(size)` in `view_noop`.

Example: 
```python
import torch

@torch.compile()
def f(x):
    batch_size = x.shape[0]
    x = x.transpose(1, 2) # (batch_size, 2, 3)
    x = x.reshape(batch_size, 2, 3) # noop
    return x

x = torch.randn((2,3,2))
f(x)

x = torch.randn((4,3,2))
f(x)
```

Before:
image

After:
image

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: test/distributed/tensor/parallel/test_micro_pipeline_tp.py [hunk 1]
```diff
@@ -123,7 +123,7 @@ def func(
         self.assertEqual(all_gathers[2].gather_dim, 0)
         self.assertEqual(
             all_gathers[2].res_node.target,
-            torch.ops.aten.view.dtype,
+            torch.ops._c10d_functional.wait_tensor.default,
```
[the remark this answers is not in the document]
Comment (BoyuanFeng) [reply test/distributed/tensor/parallel/test_micro_pipeline_tp.py:126]: offline discussed with @weifengpy and the change makes sense

File: torch/_inductor/fx_passes/fuse_attention.py [hunk 2]
```diff
@@ -641,6 +641,8 @@ def _get_sfdp_patterns():
     # attn_mask
     b_inp = functools.partial(torch.empty, (1, 1, 8, 8), device=device)
     m_inp = functools.partial(torch.empty, (2, 1, 1, 4), device=device)
+    # need 2d attn_mask to generate patterns with view op
+    m_inp_2d = functools.partial(torch.empty, (2, 4), device=device)
```
Comment (BoyuanFeng) [inline torch/_inductor/fx_passes/fuse_attention.py:645]: #113004 added `_sfdp_pattern_15`, which takes a 4d m_inp (shape: (2,1,1,4)) as `attn_mask` and calls `(attn_mask == 0).view((bs, 1, 1, k_len))`. After this pr, the view op will be noop eliminated so the search pattern will not have that view op. In `_test_sdpa_rewriter_15`, the graph still takes a 2d mask so the view op will not be noop eliminated. This difference prevents pattern match.

The change here still passes a 2d attn_mask so the view op is not noop eliminated in the search pattern.

cc @Valentine233 @jgong5 @leslie-fang-intel

Comment (Valentine233) [reply torch/_inductor/fx_passes/fuse_attention.py:645]: LGTM. Thanks!

File: torch/_inductor/fx_passes/post_grad.py [hunk 3]
```diff
@@ -882,9 +882,14 @@ def cat_noop(inputs, dim=0):
     return len(inputs) == 1
 
 
-@register_noop_decomp(aten.view)
-def view_noop(arg, size):
-    return arg.shape == size
+@register_noop_decomp(aten.view.default)
+def view_default_noop(arg, size):
+    return arg.shape == tuple(size)
+
```
Comment (eellison) [inline torch/_inductor/fx_passes/post_grad.py:888]: These should use `statically_known_true` here. We should not add a guard on arg[0] == size[0] if then arg[1] != size[1]. 

would be nice to have recursive api for this. 

but just 

`return len(arg.shape) == len(size) and all(statically_known_true(a == b) for zip(arg.shape, size))`

works for now.

cc @laithsakka

**Review by eellison — APPROVED** [review]
Looks good

File: torch/_inductor/fx_passes/post_grad.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/fx_passes/post_grad.py:888]: apparently `statically_known_true(sym_eq(ls1,ls2))` does this - thx u @laithsakka

File: torch/_inductor/fx_passes/post_grad.py [hunk 4]
```diff
@@ -898,9 +898,14 @@ def cat_noop(inputs, dim=0):
     return len(inputs) == 1
 
 
-@register_noop_decomp(aten.view)
-def view_noop(arg, size):
-    return arg.shape == size
+@register_noop_decomp(aten.view.default)
+def view_default_noop(arg, size):
+    return statically_known_true(sym_eq(arg.shape, tuple(size)))
```
Comment (BoyuanFeng) [inline torch/_inductor/fx_passes/post_grad.py:903]: arg.shape is a tuple but not a list. [`sym_eq`](https://github.com/pytorch/pytorch/pull/113159#discussion_r1385323774) returns False if one input is tuple and another input is list.

Comment (zou3519) [reply torch/_inductor/fx_passes/post_grad.py:903]: Did this used to specialize on size?
